static/app/components/modals/inviteMembersModal/useInviteModal.tsx

```diff
@@ -0,0 +1,232 @@
+import {useCallback, useEffect, useMemo, useRef, useState} from 'react';
+
+import type {
+  InviteRow,
+  NormalizedInvite,
+} from 'sentry/components/modals/inviteMembersModal/types';
+import {t} from 'sentry/locale';
+import type {Member, Organization} from 'sentry/types';
+import {trackAnalytics} from 'sentry/utils/analytics';
+import {uniqueId} from 'sentry/utils/guid';
+import {useApiQuery} from 'sentry/utils/queryClient';
+import useApi from 'sentry/utils/useApi';
+import {useIsMountedRef} from 'sentry/utils/useIsMountedRef';
+
+interface Props {
+  organization: Organization;
+  initialData?: Partial<InviteRow>[];
+  source?: string;
+}
+function defaultInvite(): InviteRow {
+  return {
+    emails: new Set<string>(),
+    teams: new Set<string>(),
+    role: 'member',
+  };
+}
+
+function useLogInviteModalOpened({
+  organization,
+  sessionId,
+  source,
+}: {
+  organization: Organization;
+  sessionId: string;
+  source: string | undefined;
+}) {
+  useEffect(() => {
+    trackAnalytics('invite_modal.opened', {
+      organization,
+      modal_session: sessionId,
+      can_invite: organization.access?.includes('member:write'),
+      source,
+    });
+  }, [organization, sessionId, source]);
+}
+
+export default function useInviteModal({organization, initialData, source}: Props) {
+  const api = useApi();
+  const willInvite = organization.access?.includes('member:write');
+  const isMounted = useIsMountedRef();
+
+  /**
+   * Used for analytics tracking of the modals usage.
+   */
+  const sessionId = useRef(uniqueId());
+  useLogInviteModalOpened({organization, sessionId: sessionId.current, source});
+
+  const memberResult = useApiQuery<Member>(
+    [`/organizations/${organization.slug}/members/me/`],
+    {
+      staleTime: 0,
+    }
+  );
+
+  const [state, setState] = useState(() => {
+    return {
+      pendingInvites: initialData
+        ? initialData.map(initial => ({
+            ...defaultInvite(),
+            ...initial,
+          }))
+        : [defaultInvite()],
+      inviteStatus: {},
```

Comment:
probably need to define a type for the whole state or at least for this object since i think it will be type `{}`

Reply:
i'll followup on it, using a reducer will change the game here for the better